### PR TITLE
[Spark] Feature delegate session catalog and spark test context rebuild

### PR DESCRIPTION
**Pull request by HuangFru**
## Why are the changes needed?

Need a parameter to control whether delegate hive table to arctic table to avoid AMS is unavailable or arctic code's bad influences for normal spark task.

spark3's test context should be rebuild to be more Reasonable

## Brief change log

  - *add logic to delegate mode for session catalog*
  - *add a spark config parameter to control whether delegate hive table to arctic table*
  - *add tests for this parameter*
  - *rebuild spark3 test context*
  - *add arctic v0.3.1 documentation for spark*

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: hive/src/test/java/com/netease/arctic/hive/HMSMockServer.java [hunk 1]
```diff
@@ -51,7 +51,7 @@ public class HMSMockServer {
   public static final Logger LOG = LoggerFactory.getLogger(HMSMockServer.class);
 
   private static final String DEFAULT_DATABASE_NAME = "default";
-  private static final int DEFAULT_POOL_SIZE = 5;
+  private static final int DEFAULT_POOL_SIZE = 500;
```
Comment (wangtaohz) [inline hive/src/test/java/com/netease/arctic/hive/HMSMockServer.java:54]: why change default pool to 500?

File: spark/v3.1/spark/src/test/java/com/netease/arctic/spark/ArcticSparkCatalogTestGroup.java [hunk 2]
```diff
@@ -6,9 +6,9 @@
  * to you under the Apache License, Version 2.0 (the
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
- *
+ *  *
```
Comment (zhoujinsong) [inline spark/v3.1/spark/src/test/java/com/netease/arctic/spark/ArcticSparkCatalogTestGroup.java:9]: Wrong licence format.

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/ArcticSparkExtensions.scala [hunk 3]
```diff
@@ -47,9 +47,9 @@ class ArcticSparkExtensions extends (SparkSessionExtensions => Unit) {
     extensions.injectOptimizerRule { spark => RewriteUpdate(spark) }
     extensions.injectOptimizerRule { spark => RewriteMergeInto(spark) }
     // arctic optimizer rules
-    extensions.injectOptimizerRule { spark => RewriteAppendArcticTable(spark) }
-    extensions.injectOptimizerRule { spark => RewriteDeleteFromArcticTable(spark) }
-    extensions.injectOptimizerRule { spark => RewriteUpdateArcticTable(spark) }
+//    extensions.injectOptimizerRule { spark => RewriteAppendArcticTable(spark) }
```
Comment (zhoujinsong) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/ArcticSparkExtensions.scala:50]: Why remove these rules?

**Review by zhoujinsong — CHANGES_REQUESTED** [review]

File: spark/v3.1/spark/src/test/java/com/netease/arctic/spark/ArcticSparkCatalogTestGroup.java (hunk 2 above)
Comment (HuangFru) [reply spark/v3.1/spark/src/test/java/com/netease/arctic/spark/ArcticSparkCatalogTestGroup.java:9]: i will change it

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/ArcticSparkExtensions.scala (hunk 3 above)
Comment (HuangFru) [reply spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/ArcticSparkExtensions.scala:50]: next version's function, now it does not work

File: spark/v3.1/spark/src/main/java/com/netease/arctic/spark/ArcticSparkSessionCatalog.java [hunk 4]
```diff
@@ -163,8 +170,8 @@ public StagedTable stageCreateOrReplace(
 
   @Override
   public Table alterTable(Identifier ident, TableChange... changes) throws NoSuchTableException {
-    if (arcticCatalog.tableExists(ident)) {
-      return arcticCatalog.alterTable(ident, changes);
+    if (isDelegateEnable() && arcticCatalog.tableExists(ident)) {
```
Comment (zhoujinsong) [inline spark/v3.1/spark/src/main/java/com/netease/arctic/spark/ArcticSparkSessionCatalog.java:173]: use `getArcticCatalog().tableExists` instead

File: spark/v3.1/spark/src/main/java/com/netease/arctic/spark/ArcticSparkSessionCatalog.java [hunk 5]
```diff
@@ -174,15 +181,19 @@ public Table alterTable(Identifier ident, TableChange... changes) throws NoSuchT
...
+    }
   }
 
   @Override
   public void renameTable(Identifier from, Identifier to) throws NoSuchTableException, TableAlreadyExistsException {
     // rename is not supported by HadoopCatalog. to avoid UnsupportedOperationException for session catalog tables,
     // check table existence first to ensure that the table belongs to the Iceberg catalog.
-    if (arcticCatalog.tableExists(from)) {
-      arcticCatalog.renameTable(from, to);
+    if (isDelegateEnable() && arcticCatalog.tableExists(from)) {
```
Comment (zhoujinsong) [inline spark/v3.1/spark/src/main/java/com/netease/arctic/spark/ArcticSparkSessionCatalog.java:195]: use `getArcticCatalog().tableExists` instead

**Review by zhoujinsong — CHANGES_REQUESTED** [review]

**Review by zhoujinsong — APPROVED** [review]

**Pull request merged**